---
title: "Plots and further analyis from output of PhyloAcc"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T)
```


This Rnotebook will guide you to generate the sequence alignment and substitution rates shift pattern for each DNA element, as well as average acceleration and conservation state per branch over all input elements. It will need:

1. A rooted phylogenetic tree in Newick format
2. Output files from PhyloAcc: *xxx_species_names.txt*, *xxx_elem_lik.txt*, *xxx_rate_postZ_xxx*
3. (To plot sequence alignment) Concatenated alignment file (*.fasta) of all elements same as input to PhyloAcc
4. (To plot sequence alignment) A bed file specify the position of each element within the concatenated alignment same as input to PhyloAcc
5. All the R functions for this notebook are in *drawAlign_function.R*.

```{r init, include= FALSE}
setwd("/Users/hzr/GitHub/PhyloAcc/R")
source("drawAlign_function.R")
```

### Read in tree data
Inputs to ```prepare_data``` are:

1. **tree_path** takes a file (*.mod*) contains a rooted phylogenetic tree in Newick format including the topology and branch lengthes, which could be output from phyloFit (PHAST); 

2. **species_name** takes a file output by PhyloAcc (*xxx_species_names.txt*) containing the species names in the same order as the columns in *xxx_rate_postZ_xxx* files. 

3. (Optional) a file with common name of the species to be shown in the plots: the argument **common_name** is optional which takes a file with abbreviation of species name appeared in the  sequences and tree files in the first column and species common name shown on the plot in the second column, and extra columns (not used in the scripts) such as full species names. 

The function ```prepare_dat``` will output a list containing the tree and names etc. as input to ```plotZPost``` and ```plotAlign``` to generate plots.

```{r treeData, cache=F}
treeData <- prepare_data(tree_path = "../Data/ratite/neut_ver3_final.named.mod", species_name = "../Data/ratite/species_names.txt", common_name = "../Data/ratite/birdname2.txt")
```

### Generate evolutionary pattern and sequence alignment for one element from PhyloAcc outputs 

```plotZPost``` function will show the shift pattern of substitution rates on the phylogengy. 
It will need:

1. *xxx_elem_lik.txt*: output by PhyloAcc constaining log-likelihood under different models for each element. 

2. *xxx_rate_postZ_xxx*: output by PhyloAcc having the posterior median of substitution rates under each conservation state (Z) and posterior of Z on each branch for different models. In the following example, we will use result from *xxx_rate_postZ_1.txt* and plot the conservation state under full model.

```{r results, echo=TRUE,cache=F}
#### read in BF scores as well as marginal log likelihood under null, accelerated and full model #### 
score <- read.table("../Simulation_ratite/result_phyloAcc/simu_500_200_npc_2-6_elem_lik.txt", header=T)
## compute BF2
score$BF2 <- -score$loglik_all + score$loglik_RES
## order score by BF1
score <- score[order(-score$log_ratio),]

#### read in posteriors of substitution rates and latent conservation states ####
postZ <- read.table("../Simulation_ratite/result_phyloAcc/simu_500_200_npc_2-6_rate_postZ_1.txt", header=T, check.names = F)  
```

Select an ratite-accelerated element (e.g. BF2 > 0 and large BF1) and plot: 
```{r plotZPost, fig.height = 8}
sel <- which(score$BF2 > 0)[1] # select the element with largest BF1 and BF2 > 0
lk = score[sel, ]
k = score[sel, 1] # get the No. of the selected element
targets = c("strCam","rhePen","rheAme","casCas","droNov","aptRow","aptHaa","aptOwe","anoDid") # target species
Z = unlist(postZ[postZ$No. == k, ]) # get the posteriors of conservation states
tit = paste("BF1:", round(lk$log_ratio), "BF2:",round(lk$BF2), "  ") # use BF scores and posterior substitution rates as title
plotZPost(Z, treeData, target_species=targets, tit=tit, offset=6) # offset= 6 indicates the posterior of Z start from 7th column
```

```plotAlign``` function will show the sequence alignment for an element as a heatmap. 
It will need 1) a bed file and 2) sequence alignments. 

In the example bed file for simulation, the 1th column is element name; the 2nd and 3rd columns are start and end positions in the alignment for each element; the 6th and 7th columns are conserved and accelerated rates to generate the DNA sequences for that element.
```{r readAlign, fig.height = 8}
bed <- read.delim("../Simulation_ratite/simu_500_200_diffr_2-6.bed", header=F)
fasta <- read.alignment(file = "../Simulation_ratite/simu_500_200_diffr_2-6.fasta", format = "fasta")  
align <- as.matrix(fasta)
align <- align[treeData$tree$tip.label,]  # reorder species in the alignment to be the same as tips of the tree. The name of the species in the alignment file has to the same as in the tree!
```

To plot the substitutions (as well as indels and unknown base pairs) of the kth element across species, 
```{r plotAlign, fig.height = 8,cache=FALSE}
plotAlign(k, align, bed, treeData, target_species=targets)
```

### Expected number of independent acceleration and accelerated species per element
Next, we could compute expected number of independent acceleration and accelerated species within target species. We will first select all the ratite accelerated elements (in this simulation, all elements are ratite accelerated and expectedly, the algorithm selects almost all the simulated elements), and then compute and get the histogram of expected number of independent acceleration within clades of target species as well as accelerated target species for each element.
```{r loss}
sel <- score$No.[which(score$log_ratio>10 & score$BF2 > 0)] # select the ratites accelerated elements
internal <- getInternals(treeData$tree, targets) # get common ancestors of target species that might be accelerated
topZ = postZ[postZ$No. %in% sel, seq(10, ncol(postZ), by=4)] # get the posterior of Z==2 (being accelerated) for selected elements
colnames(topZ) <- sapply(colnames(topZ), function(x) strsplit(x, "_")[[1]][1])

acc = rowSums(topZ[,targets]); # expected number of accelerated species within target species
loss = rowSums(topZ[,targets]) - rowSums(topZ[,internal]) # expected independent acceleration happened
par(mfrow=c(1,2), mar=c(4, 4, 2, 2), mgp = c(2, .5, 0),cex.lab = 1.5,cex.axis=1.2,lwd=2)
hist(acc, main="", xlab ="# of accelerated species"  ); 
hist(loss, main="", xlab= "# independent acceleration")
```

In this case, as the simulated configuration is that all ratites branches accelerated, so there are three independent loss (i.e. one at ancestor of kiwi-Rhea clades, one for Moa and one for Ostrich) and all 9 target species are accelerated. From the histogram, we could see that the model correctly identifies 3 losses and 9 acceleratin branches within ratites. For real data which contains different acceleration patterns, these distributions will be more flat. To get elements that might be associated with phenotypically convergence, we could select elements with multiple independent acceleration and/or accelerated in most target species.

### Evolutionary patterns over all elements
Finally, we could get the overall posterior probability of acceleration on each branch for all the input elements. The higher means that branch is more likely to be accelerated across the whole genome (or all inputs). 
```{r overall}
plotZPost_all(treeData, topZ, targets) # treeData is from first step; topZ are ratite-accelerated elements selected in the previous section; targets are target species (here, ratites)
```

In this example, all elements are accelerated only in ratite branches, and our model detected the pattern correctly as the acceleration probability is nearly 1 for ratites (shown as red) but it is nearly 0 for non accelerated branches. But for the real data, since all patterns are mixed together, the average posterior probability of accelerated will not be concentrated at 1.

